# <img style="float: left; padding-right: 10px; width: 200px" src="https://raw.githubusercontent.com/trivikverma/researchgroup/master/assets/media/logo.png"> EPA-1316 Introduction to *Urban* Data Science 


## Assignment 1: Data Collection and Wrangling


---



# ``Instructions``

This assignment puts together what you learned in **Weeks 1-2**. You will be working with a dataset which is in the form of a spreadsheet. It may contain many different data types in the columns. All data frames contain column names, which are strings, and row indices, which are integers. In this assignment you will illustrate your knwoledge about bundling various kinds of data together to be able to do higher-level tasks.

_Note:_ Go through **labs and homeworks 00-02** before starting this assignment. 

#### 1.1 Submission

Please submit the results by Brightspace under **Assignment 01**, using a single file as example,

```text
firstname_secondname_thirdname_lastname_01.html

```

**If your file is not named in lowercase letters as mentioned above, your assignment will not be read by the script that works to compile > 200 assignments and you will miss out on the grades. I don't want that, so be exceptionally careful that you name it properly. Don't worry if you spelled your name incorrectly. I want to avoid a situation where I have 200 assignments all called assignment_01.html**

Please **do not** submit any data or files other than the ``html file``.

#### 1.2 How do you convert to HTML? 

There are 2 ways, 

1. from a running notebook, you can convert it into html by clicking on the file tab on the main menu of Jupyter Lab 
    * File &rightarrow; Export Notebooks as... &rightarrow; Export Notebook to HTML
2. go to terminal or command line and type
    * ``jupyter nbconvert --to html <notebook_name>.ipynb  ``


#### 1.3 Learning Objectives

This assignment is designed to support three different learning objectives. After completing the following exercises you will be able to:

* Explore variables in a dataset
* Manage missing data 
* Reshape data to get it in a form useful for statistical analysis 

#### 1.4 Tasks

This assignment requires you to go through five tasks in cleaning your data. 

1. Reading and Summarizing the Data.
2. Subsetting the Data. This extracts just the part of the data you want to analyse. 
3. Manage Missing Data. Some data is not available for all objects of interest (rows) or all variables for every object (columns). 
4. Shape the Data. We need to convert the data into a suitable format for analysis. 
5. Saving the Results. The results are saved for future use.

<br/>

***

# ``Task 1: Downloading the Data``

For this assignment we are going to use the World Development Indicators database as a source of data. The World Development Indicators is the primary data source for the World Bank, a financial institution that provides loans to developing nations for investment in national infrastructure. The database is comprised of data from officially recognized sources all over the world. The data consists of time series which in some cases dates back over fifty years. Nations are variously categorized into different groups in order to permit the comparative analysis of nations. 

You can download the data here as a csv file (It is intentional that I am not explicitly telling you where exactly you will find the csv file on this website):
http://data.worldbank.org/data-catalog/world-development-indicators

So after you unzip, we’ll work with the file ``WDIData.csv``, which is in a modified csv format. All the other files around it are informative and may be useful for you to do a better analyses. These extra files only provide more information on data sources of indicators used in the main file. Put the data in a convenient location on your computer or laptop, ideally in a folder called **data** which is next to this **jupyter notebook**. I recommend taking a look at the file in a text editor like _atom_ for any system or notepad++ for windows. These will also make your life easy for everything else on your computer. 

It’s a big file and it may take a while to load onto your laptop and into Python (running on the jupyter labs environment). 

The data is organized with one country and all the data for one indicator on each line. But there are many countries, and many indicators. Every indicator may have data reaching back from _1960_. These are all shown together on the same line. Because the data is replicated by country, the file is longer than it is wide. We call it “long” data. Thus, each country may be repeated on rows based on the indicator that is shown.

## ``Exercise: Downloading the Data``

**IMPORTANT** make sure your code can run independent of the machine. i.e. 
- Use relative path links instead of absolute paths. If your data folder is named C:/HelloKitty/MyGummyBears/IlovePython/WDIData.csv, then your program will not be reproducible on any other machine. Check out this very easy to follow and handy guide on [relative paths](https://www.delftstack.com/howto/python/relative-path-in-python/).
- Organise the data in a folder called `data` and run your notebook next to it organised as follows

```text
├── trivik_verma_01.ipynb
├── data
│   ├── WDIData.csv
```

- Load the `WDIData.csv` file into Python
- Explore it by looking at first and last 5 rows
- Programattically find and print information on the data,
    - number of columns in the data
    - names of the columns in the data 
    - number of rows in the data (excluding the header names)
    - how many unique regions/countries in the data
    - how many unique national indicators are in the data

<h1>1. Import of necessary libraries and data </h1>
    <br>
    I import pandas, an open source library for data science, to analyze the data. As a next step, I use the read_csv function from pandas to read the data from the csv file. The data is obtained from the World Bank and contains information on the development of countries measured by different indicators and over time, starting as early as 1960. The csv file is saved in a data folder to make this notebook reproducible.


In [4]:
#import pandas package and call it pd
import pandas as pd

In [5]:
# use panda's read_csv function to read the csv file and save it in wdi_data, which is a dataframe object
wdi_data = pd.read_csv('data/WDIData.csv')

<h1>2. Explore the data</h1>
<br>
I want to get an overview of the data to get a better understanding of the structure. I use the following pandas functions to do so:

 - head() --> provides the first five entries of the dataframe
 - tail() --> provides the last five entries of the dataframe
 - columns --> provides the name of all columns
 - shape --> provides the number of rows and columns of the dataframe
 - unique() --> provides the number of unique values in a column of the dataframe


In [6]:
#print the first five entries
print(wdi_data.head())


                  Country Name Country Code  \
0  Africa Eastern and Southern          AFE   
1  Africa Eastern and Southern          AFE   
2  Africa Eastern and Southern          AFE   
3  Africa Eastern and Southern          AFE   
4  Africa Eastern and Southern          AFE   

                                      Indicator Name     Indicator Code  1960  \
0  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS   NaN   
1  Access to clean fuels and technologies for coo...  EG.CFT.ACCS.RU.ZS   NaN   
2  Access to clean fuels and technologies for coo...  EG.CFT.ACCS.UR.ZS   NaN   
3            Access to electricity (% of population)     EG.ELC.ACCS.ZS   NaN   
4  Access to electricity, rural (% of rural popul...  EG.ELC.ACCS.RU.ZS   NaN   

   1961  1962  1963  1964  1965  ...       2013       2014       2015  \
0   NaN   NaN   NaN   NaN   NaN  ...  16.936004  17.337896  17.687093   
1   NaN   NaN   NaN   NaN   NaN  ...   6.499471   6.680066   6.859110   
2   NaN   N

In [7]:
#print the last five entries
print(wdi_data.tail())

       Country Name Country Code  \
383567     Zimbabwe          ZWE   
383568     Zimbabwe          ZWE   
383569     Zimbabwe          ZWE   
383570     Zimbabwe          ZWE   
383571     Zimbabwe          ZWE   

                                           Indicator Name     Indicator Code  \
383567  Women who believe a husband is justified in be...     SG.VAW.REFU.ZS   
383568  Women who were first married by age 15 (% of w...  SP.M15.2024.FE.ZS   
383569  Women who were first married by age 18 (% of w...  SP.M18.2024.FE.ZS   
383570  Women's share of population ages 15+ living wi...  SH.DYN.AIDS.FE.ZS   
383571  Young people (ages 15-24) newly infected with HIV     SH.HIV.INCD.YG   

        1960  1961  1962  1963  1964  1965  ...     2013     2014     2015  \
383567   NaN   NaN   NaN   NaN   NaN   NaN  ...      NaN      NaN     14.5   
383568   NaN   NaN   NaN   NaN   NaN   NaN  ...      NaN      NaN      3.7   
383569   NaN   NaN   NaN   NaN   NaN   NaN  ...      NaN     33.5   

This first structural analysis showed me that I have 67 columns, as the years are on the columns while to country indicator combination describes each row. Furthermore, I already see some missing values in the data. Moreover, it is also clear that the data does not only include countries but also regions, as the first entries are for the region "Africa Eastern and Southern".

In [8]:
#print the names of the columns to get a better understanding
print(wdi_data.columns)

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021',
       'Unnamed: 66'],
      dtype='object')


This output showed me that data was collected from 1960 to 2021.

In [9]:
# print the number of columns by using the python function len()
print(len(wdi_data.columns))

67


In [10]:
# To fully see the length of the dataframe, I print the number of rows
print(wdi_data.shape[0])

383572


In [11]:
# identify the number of unique countries
print(len(wdi_data['Country Code'].unique()))

266


In [12]:
#identify the number of unique indicators
print(len(wdi_data['Indicator Code'].unique()))

1442


I collected the following information:
- number of columns in the data: 67
- number of rows in the data: 383572
- number of unique countries/regions 266
- number of unique indicators  1442 <br>

I notice that they are way more rows than columns. Together with the information I got on the names of the columns, I can conclude that the data is in a long format. This means that each row contains information on one country and one indicator. The data is not in a wide format, where each row contains information on one country and all indicators. This is important to know, because I might have to reshape the data later on. Hence, I want to know how many unique countries and indicators are in the data. I use the unique() function to get this information. I notice that there are 266 unique countries/regions and 1442 unique indicators. 

<br/>

***

# ``Task 2: Subsetting the Data``

From now on we want a much smaller subset of this data. We have all the valid country information using the country code information in ``WDICountry.csv`` or in one of the columns of the main data itself. (Note that it is best practice to search using country codes and not real country names. Countries are known by many names by many different people and languages.) In the future, World Bank may change the datasets with new country names as the data collection efforts of orgs is not relevant to geopolitics. Hence, it is important to work with codes as opposed to names to make our analyses more reproducible across time.

The file ``WDISeries.csv`` contains a description of all the indicator variables and their names. We won’t actually use this file in the analysis, but you will find it helpful in designing your own analysis. Your objectives for this assignment is to select **4-7 variables** for further exploratory statistics (more information later in exercises). 

For example, I can show you what I did, 

```text
My hypothesis
I’d like to examine world broadband access. For that reason I chose a broadband account variable. The data is organized by country. I want to control for the wealth, population, and land area of the country. I also have a hypothesis that more urban countries are more likely to have good broadband services. There are economies of scale when providing services to a large city. 

I hypothesize that larger countries have lesser access, since it is expensive to provide access over larger areas. On the contrary, countries with a lot of urban land area can take advantage of economies of scale resulting in relatively more broadband users concentrated in smaller zones within the country. We also hypothesize that wealthier countries have better broadband access, since there is a larger market to provide the newest services. A final variable which we add is rail lines. I hypothesize that broadband lines can take advantage of existing infrastructure right-of-ways, of which rail is a surrogate measure. Furthermore, the presence of rail lines may indicate other factors including a geography which is conclusive to physical development, and favourable institutional factors which promote high technology development. 

My choice of variables were, 

| Variable Name                 | Variable Code     |
| ----------------------------- | ----------------- |
| Fixed broadband subscriptions | IT.NET.BBND       |
| GDP (current US$)             | NY.GDP.MKTP.CD    |
| Population, total             | SP.POP.TOTL       |
| Land area (sq. km)            | AG.LND.TOTL.K2    |
| Urban land area (sq. km)      | AG.LND.TOTL.UR.K2 |
| Rail lines (total route-km)   | IS.RRS.TOTL.KM    |

```

Recall that the data is organized with countries and variables on the rows, and years on the column. Using this table as a guide, I can now extract only those rows which contain these variable names, and throw out the great many other variables that I will not need. You are expected to do the same further down in the exercise. 

For now let’s set aside the added complexity of time series and dynamics. Our task is to select just one year with a lot of data for most countries.

## ``Exercise: Subsetting the Data``

- state your hypothesis in a markdown cell as I showed in the example above (there is no single right hypothesis, you are free to make a **reasonable** choice for this task)
- find the variables of interest for your hypothesis and mention them in the markdown cell (4-7 variables)
- your dataframe would have greatly reduced in size and looks neater, show us what it looks like now using head() or something similar
    - show some statistics like number or rows, columns, names of variables and unique countries etc.
- you’ll see that your data contains values for many years of data, or perhaps NA (“not applicable”), if the country has failed to report its findings. 
- For now let’s set aside the added complexity of time series and dynamics. Our task is to select just one year with a lot of data for most countries.
    - choose one year/column that you want to work with and drop the rest of the years. 

You can count the columns manually, but in a large data set like this it is accurate and convenient to let python calculate this for us. Get the index of relevant columns and store them in a variable. 

- when you do this for your own variables, you also will want to experiment to see which year you want to use. You might also choose to drop off some of your initial variable choices if they are poorly collected. 
- subset the data by creating a new dataframe only with ``your variables`` `[v1, v2, v3...]`

<hr>
<h3>My hypothesis</h3>

I like to analyze the a population health index. Therefore, I choose life expectancy as a proxy as life expectancy provides me with the mortality over a lifespan and is recorded for a large number of countries and regions.
I hypothesize that richer countries have a higher life expectancy as they have more resources and better infrastructure to provide health services. Hence, I include the country's wealth as a variable. I look at the UHC service coverage index, which states a score from 0 to 100 for essential health services (based on tracer interventions that include reproductive, maternal, newborn and child health, infectious diseases, noncommunicable diseases and service capacity and access). THe higher the score the higher is the coverage for essential health services. Furthermore, I argue that countries with a higher disparity between incomes, have a lower life expectancy as the country might have a high level of wealth, yet it is focused on a few while the larger part of the population will have a lower coverage. Hence, I include the Gini index as a variable, as it measures the distribution of income, 0 meaning perfect equality and 100 perfect inequality.
 I also hypothesize that countries with a higher urban population have a higher coverage as they have better infrastructure and more resources. Hence, I include the urban population as a variable. I also hypothesize that countries with a higher literacy rate have a higher coverage as they are more educated and hence have a better understanding of the importance of health services. Hence, I include the literacy rate as a variable. 
 I also look at the coverage of social insurance programs, as I argue that a higher coverage will lead to a higher life expectancy due to the fact, that more people will have access to health services and can also afford necessary but expensive treatments.
My choice of variables were, 
<br>

| Variable Name                                             | Variable Code     |
| -----------------------------                             | ----------------- |
| life expectancy                                           | SP.DYN.LE00.IN    |
| UHC service coverage index                                | SH.UHC.SRVS.CV.XD       |
| GDP (current US$)                                         | NY.GDP.MKTP.CD    |
| Gini coefficent                                           | SI.POV.GINI     |
| Coverage of social insurance programs (%)                 | per_si_allsi.cov_pop_tot |
|Literacy rate, adult total (% of people ages 15 and above) | SE.ADT.LITR.ZS |
| Hospital Beds (Per 1,000 People)                          | SH.MED.BEDS.ZS    |


<h3>Creating the subset</h3>

To obtain my subset of the original dataset, I use a python list that contains the codes for the selected indicators. I choose the indicator codes as the World Bank might change the names of the indicators over time.
Afterwards, I use this list of indicator codes together with pandas isin() function to create a subset which I save in the variable wdi_data_subset.

In [13]:
#List of selected indicators
selected_variables = ['SP.DYN.LE00.IN','SH.UHC.SRVS.CV.XD','NY.GDP.MKTP.CD','SI.POV.GINI','per_si_allsi.cov_pop_tot','SE.ADT.LITR.ZS','SH.MED.BEDS.ZS']
#subset of the dataframe with only the selected indicators
wdi_data_subset = wdi_data[wdi_data['Indicator Code'].isin(selected_variables)]


<h3>Overview on subset</h3>

To quickly check if the subset was created correctly, I use the head() function to display the first 5 rows of the subset.

In [14]:
#print the first five rows of the subset with head()
print(wdi_data_subset.head())

                    Country Name Country Code  \
243  Africa Eastern and Southern          AFE   
467  Africa Eastern and Southern          AFE   
491  Africa Eastern and Southern          AFE   
558  Africa Eastern and Southern          AFE   
697  Africa Eastern and Southern          AFE   

                                        Indicator Name  \
243  Coverage of social insurance programs (% of po...   
467                                  GDP (current US$)   
491                                         Gini index   
558                   Hospital beds (per 1,000 people)   
697            Life expectancy at birth, total (years)   

               Indicator Code          1960          1961          1962  \
243  per_si_allsi.cov_pop_tot           NaN           NaN           NaN   
467            NY.GDP.MKTP.CD  2.129059e+10  2.180847e+10  2.370702e+10   
491               SI.POV.GINI           NaN           NaN           NaN   
558            SH.MED.BEDS.ZS  1.959677e+00           Na

I see that the subset still contains 67 columns and also my selected indicators, as expected.

Furthermore, I want to get a quick overview of the new dataframe and thus, use pandas info() function to display the number of rows and columns, the column names and the data types of the columns, as well as the non-null value count for each column.

In [15]:
print(wdi_data_subset.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1862 entries, 243 to 383516
Data columns (total 67 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country Name    1862 non-null   object 
 1   Country Code    1862 non-null   object 
 2   Indicator Name  1862 non-null   object 
 3   Indicator Code  1862 non-null   object 
 4   1960            534 non-null    float64
 5   1961            375 non-null    float64
 6   1962            377 non-null    float64
 7   1963            376 non-null    float64
 8   1964            376 non-null    float64
 9   1965            394 non-null    float64
 10  1966            394 non-null    float64
 11  1967            399 non-null    float64
 12  1968            403 non-null    float64
 13  1969            404 non-null    float64
 14  1970            587 non-null    float64
 15  1971            420 non-null    float64
 16  1972            422 non-null    float64
 17  1973            423 non-null 

<h3>Selection of specific year</h3>

As this analysis will not analyze the data for each country over time, I want to select only one year. To keep the highest level of quality for my analysis, I want to find out the year with the highest number of provided values. To do so, I use the pandas count() function to count the number of non-null values for each column. I then use the idxmax() function to find the column with the highest number of non-null values. I save the result in the variable year_with_most_values. As I want to exclude the first columns that are specific to countries and regions, I use the iloc[] function and look only from the 5th column onwards.

In [16]:
#Year with most entries, only included year columns with iloc
year_with_most_values = wdi_data_subset.iloc[:,5:].count().idxmax()
print(year_with_most_values)

2010


As I have identified the year with the most values, in this case 2010, I want to create a new dataframe that only contains this year. To do so, I use the pandas loc[] function to select all rows and only the columns on the country, indicator, and selected year. I save the result in the variable wdi_data_subset_for_selected_year.

In [17]:
#list with columns that I will need in the new dataframe
relevant_columns = ['Country Name','Country Code','Indicator Name','Indicator Code',year_with_most_values]
#create new dataframe with only the relevant columns and all rows
wdi_data_subset_for_selected_year = wdi_data_subset.loc[:,relevant_columns]

In [18]:
#Obtain overview on new created dataset using info()
print(wdi_data_subset_for_selected_year.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1862 entries, 243 to 383516
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country Name    1862 non-null   object 
 1   Country Code    1862 non-null   object 
 2   Indicator Name  1862 non-null   object 
 3   Indicator Code  1862 non-null   object 
 4   2010            1086 non-null   float64
dtypes: float64(1), object(4)
memory usage: 87.3+ KB
None


<h3>Checking Quality of Selected Indicators</h3>

As a next step, I want to check the quality of each indicator for the selected year. I do so by counting the values for the selected year, in this case 2010. To do so, I use pandas groupby() function to group the dataframe by the indicator code and then use the count() function to count the number of non-null values for each indicator. I print the result to the console.

In [19]:
# display count of values for each indicator
print(wdi_data_subset_for_selected_year.groupby(['Indicator Code','Indicator Name']).count())

                                                                             Country Name  \
Indicator Code           Indicator Name                                                     
NY.GDP.MKTP.CD           GDP (current US$)                                            266   
SE.ADT.LITR.ZS           Literacy rate, adult total (% of people ages 15...           266   
SH.MED.BEDS.ZS           Hospital beds (per 1,000 people)                             266   
SH.UHC.SRVS.CV.XD        UHC service coverage index                                   266   
SI.POV.GINI              Gini index                                                   266   
SP.DYN.LE00.IN           Life expectancy at birth, total (years)                      266   
per_si_allsi.cov_pop_tot Coverage of social insurance programs (% of pop...           266   

                                                                             Country Code  \
Indicator Code           Indicator Name                              

<br/>

***

# ``Task 4: Reshape the Data``

As you may have noticed from your outputs above, the data is still not  in a form which is suitable for statistical analysis. Every row is a a combination of a country, a few variables, and a year. We’d like each row instead to be a country, and for there to be many columns according to the variables involved. 

The data is stored with one country and one variable by year. That’s long data. We want to convert it so each row is a case, and that case is a country. Then each column can store the variables for that country. That’s wide data. Our objectives in this section is to convert from one format of the data to the other. For the purposes of this assignment we’re not going to handle time series data, even though the World Development Indicators data often has many years of time history collected for each of the nations. That's why I asked you to select a particular year only.

You might ask why the original data was even stored in this manner. The most efficient means of storing data is to store everything once and then not repeat it. So for instance each element of this data set might be a combination of a country, variable and year. Any additional information, like the full and official name of the country, could be stored in a supplementary table and consulted only at need. 

That’s the most efficient way. But every user and application is slightly different. As noted above, typically what we need for statistical analysis is a single case on each row, and a set of variables in the columns. Our case is a country, and our variables are things like GDP and population as described above in my example choice of variables. This involves some restructuring of the data which we clearly don’t want to do by hand. Pandas is your best friend here. 

Reshaping data is a two-step process of melting and pivoting the data. Melting the data involves describing which data are indicators ("id") and which are variables for retrieval (“measure”). In this case your data may already be in melted form (long form). Pivoting then involves actually reshaping the data into the needed format. In this step, you have to reshape the data from long to wide format.
 
Pivoting the data involves specifying what data is on the rows and on the columns. Hint: functions melt and pivot offered by ``numpy`` library in python. For our analyses we want “Country.Code” to be on the rows, and to have all 4-7 variables as columns, where the value of each cell is the value taken from the column year that you chose at the subsetting step.



## ``Exercise: Reshape the Data``

- Examine the dimensions of the new pivoted data that you have created. Show it to us using head or print commands.
- Then rename all column names to something better and useful, by replacing codes with their names or shorthand names (ex. AG.LND.TOTL.UR.K2 ---> Urban Land Area).  
- Sort the data by putting higher values for one indicator of your choice go first. If there are overlapping values, try to put chronological countries go first.

<hr>
<h3>Reshape the Data</h3>

As the next step, I decided to reshape my data first, before I look at any missing data. Having the indicators on the columns will simplify the process of looking for missing data. 

In [32]:
# create the pivoted dataframe bringing the indicator codes to the columns
wdi_data_subset_2010_pivot = (wdi_data_subset_for_selected_year.pivot(index='Country Code', columns='Indicator Code', values=year_with_most_values))
print(wdi_data_subset_2010_pivot.head())

Indicator Code  NY.GDP.MKTP.CD  SE.ADT.LITR.ZS  SH.MED.BEDS.ZS  \
Country Code                                                     
ABW               2.453631e+09       96.822639             NaN   
AFE               8.604783e+11             NaN             NaN   
AFG               1.585668e+10             NaN            0.43   
AFW               5.915958e+11             NaN             NaN   
AGO               8.169956e+10             NaN             NaN   

Indicator Code  SH.UHC.SRVS.CV.XD  SI.POV.GINI  SP.DYN.LE00.IN  \
Country Code                                                     
ABW                           NaN          NaN       75.017000   
AFE                           NaN          NaN       58.470697   
AFG                          28.0          NaN       61.028000   
AFW                           NaN          NaN       54.144307   
AGO                          32.0          NaN       55.350000   

Indicator Code  per_si_allsi.cov_pop_tot  
Country Code                   

In [31]:
wdi_data_subset_2010_pivot['SI.POV.GINI']


Country Code
ABW     NaN
AFE     NaN
AFG     NaN
AFW     NaN
AGO     NaN
       ... 
XKX    33.3
YEM     NaN
ZAF    63.4
ZMB    55.6
ZWE     NaN
Name: SI.POV.GINI, Length: 266, dtype: float64

In [21]:
indicator_code_to_text = {
    'SP.DYN.LE00.IN': 'Life expectancy at birth, total (years)',
    'SH.UHC.SRVS.CV.XD': 'UHC service coverage index',
    'NY.GDP.MKTP.CD': 'GDP (current US$)',
    'SI.POV.GINI': 'GINI index (World Bank estimate)',
    'SE.ADT.LITR.ZS': 'Literacy rate, adult total (% of people ages 15 and above)',
    'SH.MED.BEDS.ZS': 'Hospital beds (per 1,000 people)'
}
wdi_data_subset_2010_pivot = (wdi_data_subset_2010_pivot.rename(columns=indicator_code_to_text))
print(wdi_data_subset_2010_pivot.info())

<class 'pandas.core.frame.DataFrame'>
Index: 266 entries, ABW to ZWE
Data columns (total 7 columns):
 #   Column                                                      Non-Null Count  Dtype  
---  ------                                                      --------------  -----  
 0   GDP (current US$)                                           256 non-null    float64
 1   Literacy rate, adult total (% of people ages 15 and above)  91 non-null     float64
 2   Hospital beds (per 1,000 people)                            171 non-null    float64
 3   UHC service coverage index                                  204 non-null    float64
 4   GINI index (World Bank estimate)                            84 non-null     float64
 5   Life expectancy at birth, total (years)                     248 non-null    float64
 6   per_si_allsi.cov_pop_tot                                    32 non-null     float64
dtypes: float64(7)
memory usage: 16.6+ KB
None


<br/>

***

# ``Task 3: Manage Missing Data``

There is a lot of missing data. If you make the year on which you search too recent, many countries have not been able to report their data. If you make the year too long ago, the practice of administrative data collection had not yet taken hold. Countries did not know that collecting data would be a good thing; furthermore they have yet to back-fill their records. 

Why is data not available or missing in this dataset?
The data availability for urbanisation is especially limited. The urbanisation variable in particular ``AG.LND.TOTL.UR.K2`` is only available per decade. The most recent populated data for this example variable is therefore 2010, in variable ``2010``.
It’s quite possible that the World Bank is constructing estimates of urbanisation out of complex data sources such as satellite imagery. Regardless, it appears expensive to compute, and is therefore only offered every decade. 

We’ve got a number of ways in general of dealing with missing data. These involve

1. Dropping off cases (or rows) in the data with any missing variables
2. Excluding variables in the data with any missing data 
3. Selectively choosing indicators with only a limited amount of missing data
4. Replacing missing variables with averages, or other representative values
5. Creating a separate model to predict missing data

In this assignment we are going to use a number of these strategies. We can certainly be dropping off cases (strategy one). I am loathe to drop off whole indicators. But we can, for example, choose a year for the indicator where most of the data is available (strategy three).

Building a separate model to impute missing data, is often a good idea. But that requires a first working model before we even consider building a missing data model (and we haven't got there yet in this course); the working model and the missing data model are often constructed together. Note also that there are packages in Python which will construct a model of your data, and then impute missing values for you. You may or may not find these functions and packaging for modelling your data to be fully appropriate. Therefore treat these missing data models very seriously, and not as a black box. Models of missing data are as important, and deserve just as much care and caution as any other statistical model.  

In the next section I discuss some specifics about how the data is currently formatted, and how we would like to have it formatted for analysis purposes. 


## ``Exercise: Manage Missing Data``

We’ve got a number of ways in general of dealing with missing data. These involve

1. Dropping off cases (or rows) in the data with any missing variables
2. Excluding variables in the data with any missing data 
3. Selectively choosing indicators with only a limited amount of missing data
4. Replacing missing variables with averages, or other representative values
5. Creating a separate model to predict missing data

- Count the missing values in each column
- Manage the missing values (delete or replace values or leave them as they are) and briefly explain your choice for each column using comments or markdown text


In [22]:
# your code here
# use many cells if you like to structure your code well
print(wdi_data_subset_for_selected_year.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1862 entries, 243 to 383516
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country Name    1862 non-null   object 
 1   Country Code    1862 non-null   object 
 2   Indicator Name  1862 non-null   object 
 3   Indicator Code  1862 non-null   object 
 4   2010            1086 non-null   float64
dtypes: float64(1), object(4)
memory usage: 87.3+ KB
None


In [23]:
wdi_data_subset_2010 = wdi_data_subset_for_selected_year.dropna()
print(wdi_data_subset_for_selected_year.info())

#drop them because Gini coefficent cannot be set to 0 - it is a measure of inequality

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1862 entries, 243 to 383516
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country Name    1862 non-null   object 
 1   Country Code    1862 non-null   object 
 2   Indicator Name  1862 non-null   object 
 3   Indicator Code  1862 non-null   object 
 4   2010            1086 non-null   float64
dtypes: float64(1), object(4)
memory usage: 87.3+ KB
None


<br/>

***

# ``Task 5: Saving the Results``

_Note:_ We do not need this file but we expect that if you learn how to save your data, it will be very useful in the future, as you do not need to run the script to clean your data again. 

## ``Exercise: Saving the Results``
- Save the cleaned dataframe as 'assignment-01-cleaned.csv' in data folder


In [24]:
wdi_data_subset_2010_pivot.to_csv('data/WDIData_2010.csv')